enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,308 @@
+encapsulated in a new SLO.
+
+Ensure that this new SLO has minimal impact on clusters not running on
+the `baremetal` platform.
+
+### Non-Goals
+
+### Proposal
+
+Recognizing that bare metal support warrants the creation of a new
+"subsystem":
+
+1. Create a new, OpenShift-specific project called the Bare Metal
+   Operator.
+2. This project should implement a new SLO which is responsible for
+   installing bare metal specific CRDs and running the BareMetalHost
+   controller and related Metal3 components.
+3. The BMO should meet all of the standard expectations of an SLO,
+   including for example keeping a ClusterOperator resource updated.
+4. On infrastructure platforms other than `baremetal`, the BMO should
+   following emerging patterns for "not in use" components, for
+   example setting its cluster operator condition to `Disabled=true`,
+   `Available=true`, and `Progressing=False`.
+
+### Risks and Mitigations
+
+(FIXME)
+
+- Impact on non-baremetal platforms
+- Implementation cost and complexity
+- Upgrades
+- API stability
+- Additional bare metal specific functionality in future
+- Upstream (Metal3) implications
+
+## Design Details
+
+The Bare Metal Operator is a new Second Level Operator (SLO) whose
+operand is a controller for the BareMetalHost resource and associated
+components from the Metal3 project. The below sections covers
+different areas of the design of this new SLO.
+
+### Standard SLO Behaviors
+
+As an SLO, the BMO is expected to adhere to the standard expected
+behaviours of SLO, including:
+
+1. The BMO image should be tagged with
+   `io.openshift.release.operator=true` and contain a `/manifests`
+   directory with all of the manifests it requires the CVO to apply,
+   along with an `image-references` file listing the images referenced
+   by those manifests that need to be included in the OpenShift
+   release image.
+2. Implement a `ClusterOperator` resource, including updating its
+   status field with the `Available/Progressing/Degraded` conditions,
+   operator version number, and any `relatedObjects`.
+
+While not required initially, other common SLO patterns can be
+considered in future:
+
+1. Implement an operator configuration resource, including
+   `OperatorSpec` and `OperatorStatus` (as per in openshift/api#125)
+2. Implement cluster-level configuration under `config.openshift.io`
+   (as per openshift/api#124)
+3. Expose a `/metrics` endpoint for Prometheus to be configured to
+   scrape (via a `ServiceMontitor`) and define any relevant Prometheus
+   alert rules based on those metrics.
+
+### "Not In Use" SLO Behaviors
+
+Unlike most other SLOs, the BMO is not applicable to all cluster
+configurations. On clusters running on an infrastructure platform
+other than `baremetal` it should adhere to the emerging expected
+behaviors for "not in use" SLOs, including:
+
+1. Setting its `ClusterOperator` condition to `Disabled=true`,
+  `Available=true`, `Progressing=False` with appropriate messages.
```

Comment:
@markmc @sadasu I've been discussing this work with @romfreiman and @yevgeny-shnaidman, there is a requirement to disable the metal3 components for the baremetal platform in some cases, and I'm wondering how we'd achieve that if the operator is still coupled to the baremetal platform?

Could we perhaps add a flag to the OperatorSpec which the installer can set based on some install-config input, e.g to enable/disable the managed baremetal components?

While the metal3 components are managed via the MAO we could probably add a flag to the baremetal CR added via https://github.com/openshift/enhancements/pull/119 which conditionally applies the metal3 deployment, but it seems like such an approach may not be applicable to this new operator (and thus is likely to be rejected if proposed as a short-term solution)?

Reply:
I think, looking forward we should support not only disable but also specify different BM provides to the operator (maybe can think of 2 potential providers - ironic / assisted).

Reply:
@hardys, maybe, since this is a short-term solution we can manage without adding a new flag to baremetal CR. This CR is included in the bootstrap ignition, so we can override it, and if all the configuration fields are set to null (DHCP External, DHCP Range, IP, Interface, Network CIDR, OS Download URL), then it means that MAI should not deploy Baremetal Controller at all. So the only change that will need to be done is to change the MAO code in syncBaremetalControllers

Reply:
@hardys can we use the provisioning-configuration CR to do this? It almost works like that now. Metal3 deployment happens only when platform type is BareMetal and provisioning-configuration is available.

Reply:
> I think, looking forward we should support not only disable but also specify different BM provides to the operator (maybe can think of 2 potential providers - ironic / assisted).

@romfreiman for the different BM providers, do you want to still deploy the metal3 pod but with different containers in it? Or do you not want to deploy metal3 pod at all in the "assisted" case?

Reply:
> @hardys, maybe, since this is a short-term solution we can manage without adding a new flag to baremetal CR. This CR is included in the bootstrap ignition, so we can override it, and if all the configuration fields are set to null (DHCP External, DHCP Range, IP, Interface, Network CIDR, OS Download URL), then it means that MAI should not deploy Baremetal Controller at all. So the only change that will need to be done is to change the MAO code in syncBaremetalControllers

@yevgeny-shnaidman @hardys IMO we do not need to add an additional flag at to the CR at all. Just the absence/presence of the CR can determine operator behavior. What do you think?

Reply:
@sadasu I think that, currently, in case baremetal CR is missing, the whole reconciliation loop of MAO will exit on error, in which case it will be scheduled again and we will have log saturated with errors. IMHO it is best to avoid this scenario. What do you think about not adding a flag, but using null (empty values) for all the configuration fields of CR as a flag?

Reply:
I feel like we're getting way off-topic in this conversation.

This enhancement is about moving the current bare metal functionality out of `machine-api-operator` into `cluster-baremetal-operator` with minimal changes.

Take a look at markmc/cluster-baremetal-operator for more details - e.g. the platform type in the infrastructure resource, and the provisioning resource are the only resources that affect this operator's functionality currently. Just like in `machine-api-operator`.

Yes, we can add more functionality and configurability to the new operator. We could even add new functionality to `machine-api-operator` first and copy it across to the new operator, if we believe the new functionality should be incorporated before the new operator.

(For example - if we want to make `infrastructure.platform == "baremetal" && provisioning == null` mean that the `metal3` deployment should not be created, then we can do that in `machine-api-operator` or `cluster-baremetal-operator` or both, depending on our desired timing for releasing that change)

Reply:
Ok, I've updated the summary to make it more clear that the enhancement is about adopting functionality from MAO and future bare metal functionality would be added to CBO after we've made this move.

Reply:
> @sadasu I think that, currently, in case baremetal CR is missing, the whole reconciliation loop of MAO will exit on error, in which case it will be scheduled again and we will have log saturated with errors. IMHO it is best to avoid this scenario. What do you think about not adding a flag, but using null (empty values) for all the configuration fields of CR as a flag?

@yevgeny-shnaidman Let us take the discussion about this new work you are requesting out of this review. It is really not related to what is being discussed here.